### PR TITLE
Build bmp container into sonic-buildimage

### PR DESCRIPTION
**Pull request by FengPan-Frank**
#### Why I did it
Build bmp container into sonic-buildimage, and added relevant daemon/file handling.

##### Work item tracking
- Microsoft ADO **(number only)**:27588893

#### How I did it
Build bmp container into sonic-buildimage, and added relevant daemon/file handling.

#### How to verify it
Local build successfully and verified in lab DUT.

image

image

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

File: dockers/docker-sonic-bmp/critical_processes [hunk 1]
```diff
@@ -0,0 +1,4 @@
+group:sonic-bmp
+program:openbmpd
+program:bmpcfgd
+{%- endif %}
```
Comment (qiluo-msft) [inline dockers/docker-sonic-bmp/critical_processes:4]: >endif [](http://example.com/codeflow?start=4&length=5)

did you miss `if` statement? did you test?

File: dockers/docker-sonic-bmp/docker_init.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+udp_server_ip=$(ip -j -4 addr list lo scope host | jq -r -M '.[0].addr_info[0].local')
+hostname=$(hostname)
+# Generate the following files from templates:
+# port-to-alias name map
+sonic-cfggen -d -t /usr/share/sonic/templates/rsyslog.conf.j2 \
+    -a "{\"udp_server_ip\": \"$udp_server_ip\", \"hostname\": \"$hostname\"}" \
+    > /etc/rsyslog.conf
+
+# Make the script that waits for all interfaces to come up executable
+chmod +x /usr/bin/start.sh
```
Comment (qiluo-msft) [inline dockers/docker-sonic-bmp/docker_init.sh:16]: We could ensure `+x` during build time. Do we really need it in runtime?

File: rules/config [hunk 3]
```diff
@@ -152,6 +152,9 @@ INCLUDE_DHCP_RELAY = y
 # INCLUDE_DHCP_SERVER - build and install dhcp-server package
 INCLUDE_DHCP_SERVER ?= n
 
+# INCLUDE_BMP - build and install sonic-bmp package
```
Comment (qiluo-msft) [inline rules/config:155]: >build [](http://example.com/codeflow?start=16&length=5)

Build time, we should "include".
Runtime, you can decide by default "enable" or not.

Comment (FengPan-Frank) [reply rules/config:155]: Yes, I will turn this on as last step after bmp/bmpcfgd PR are merged.

File: dockers/docker-sonic-bmp/critical_processes (hunk 1 above)
Comment (FengPan-Frank) [reply dockers/docker-sonic-bmp/critical_processes]: fixed

File: dockers/docker-sonic-bmp/docker_init.sh (hunk 2 above)
Comment (FengPan-Frank) [reply dockers/docker-sonic-bmp/docker_init.sh]: fixed

**Comment by forrestchu** [comment]
> #### Why I did it
> Add bmp container init workflow.
> 
> ##### Work item tracking
> * Microsoft ADO **(number only)**:27588893
> 
> #### How I did it
> Add bmp container which is disabled by INCLUDE_BMP = n for now, which would cause container not build by default. This will be enabled as last step.
> 
> Add INCLUDE_BMP to indicate whether to build bmp container Add docker file for bmp, as well as agent used openbmpd/bmpcfgd. Add template file for bmp container services. Add entry for bmp to FEATURE table in config_db.
> 
> #### How to verify it
> Build image with INCLUDE_BMP = y to verify, Image should be install correctly. By config feature state bmp enabled to enable.
> 
> #### Which release branch to backport (provide reason below if selected)
> * [ ]  201811
> * [ ]  201911
> * [ ]  202006
> * [ ]  202012
> * [ ]  202106
> * [ ]  202111
> * [ ]  202205
> * [ ]  202211
> * [ ]  202305
> 
> #### Tested branch (Please provide the tested image version)
> * [ ]
> * [ ]
> 
> #### Description for the changelog
> #### Link to config_db schema for YANG module changes
> #### A picture of a cute animal (not mandatory but encouraged)

Hi Frank, which TCP port will this bmp server use by default? Is it configurable from config_db or CLI?

File: files/build_templates/init_cfg.json.j2 [hunk 4]
```diff
@@ -60,6 +60,7 @@
 {%- if include_sflow == "y" %}{% do features.append(("sflow", "disabled", true, "enabled")) %}{% endif %}
 {%- if include_macsec == "y" %}{% do features.append(("macsec", "{% if 'type' in DEVICE_METADATA['localhost'] and DEVICE_METADATA['localhost']['type'] == 'SpineRouter' and DEVICE_RUNTIME_METADATA['MACSEC_SUPPORTED'] %}enabled{% else %}disabled{% endif %}", false, "enabled")) %}{% endif %}
 {%- if include_system_gnmi == "y" %}{% do features.append(("gnmi", "enabled", true, "enabled")) %}{% endif %}
+{%- if include_bmp == "y" %}{% do features.append(("bmp", "enabled", true, "enabled")) %}{% endif %}
```
Comment (qiluo-msft) [inline files/build_templates/init_cfg.json.j2:63]: We should not change init_cfg.json.j2, in order to enable by default in image. This enablement is controlled by golden config generation service outside sonic.

**Comment by FengPan-Frank** [comment]
> ?

it's 5000, pls refer HLD section https://github.com/sonic-net/SONiC/blob/50cb55989649d5392b48b54bb7ca8263c883e162/doc/bmp/bmp.md#config-file-update

File: dockers/docker-sonic-bmp/base_image_files/redis-cli [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+DOCKER_EXEC_FLAGS="i"
+
+# Determine whether stdout is on a terminal
+if [ -t 1 ] ; then
+    DOCKER_EXEC_FLAGS+="t"
+fi
+
+docker exec -$DOCKER_EXEC_FLAGS bmp redis-cli "$@"
```
Comment (qiluo-msft) [inline dockers/docker-sonic-bmp/base_image_files/redis-cli:10]: why you use this file?

**Review by qiluo-msft — APPROVED** [review]

**Review by xincunli-sonic — APPROVED** [review]
LGTM

File: dockers/docker-sonic-bmp/base_image_files/redis-cli (hunk 5 above)
Comment (FengPan-Frank) [reply dockers/docker-sonic-bmp/base_image_files/redis-cli]: removed

**Review by qiluo-msft — APPROVED** [review]

**Review by FengPan-Frank — COMMENTED** [review]
https://github.com/sonic-net/sonic-host-services/pull/128, bmpcfgd unittest passed in sonic-host-services before.

**Review by qiluo-msft — APPROVED** [review]

File: files/image_config/rsyslog/rsyslog.d/00-sonic.conf.j2 [hunk 6]
```diff
@@ -47,6 +47,12 @@ if $msg startswith " gnmi-native" then {
     stop
 }
 
+## bmp rules
```
Comment (qiluo-msft) [inline files/image_config/rsyslog/rsyslog.d/00-sonic.conf.j2:50]: No need to add standalone logging file

Comment (FengPan-Frank) [reply files/image_config/rsyslog/rsyslog.d/00-sonic.conf.j2:50]: will raise new PR for fixing this later.

**Pull request merged**